query/physicalplan/aggregate.go

```diff
@@ -236,103 +238,110 @@ func hashInt64Array(arr *array.Int64) []uint64 {
 	return res
 }
 
-func (a *HashAggregate) Callback(ctx context.Context, r arrow.Record) error {
-	// Generates high volume of spans. Comment out if needed during development.
-	// ctx, span := a.tracer.Start(ctx, "HashAggregate/Callback")
-	// defer span.End()
-
-	groupByFields := a.groupByFields
-	groupByFieldHashes := a.groupByFieldHashes
-	groupByArrays := a.groupByArrays
-
-	defer func() {
-		groupByFields = groupByFields[:0]
-		groupByFieldHashes = groupByFieldHashes[:0]
-		groupByArrays = groupByArrays[:0]
-	}()
-
-	var columnToAggregate arrow.Array
-	aggregateFieldFound := false
-
-	for i, field := range r.Schema().Fields() {
-		for _, matcher := range a.groupByColumnMatchers {
-			if matcher.MatchColumn(field.Name) {
-				groupByFields = append(groupByFields, field)
-				groupByFieldHashes = append(groupByFieldHashes, scalar.Hash(a.hashSeed, scalar.NewStringScalar(field.Name)))
-				groupByArrays = append(groupByArrays, r.Column(i))
-			}
-		}
+func (a *HashAggregate) SetNext(next PhysicalPlan) {
+	a.next = next
 
-		if a.columnToAggregate.MatchColumn(field.Name) {
-			columnToAggregate = r.Column(i)
-			aggregateFieldFound = true
-		}
-	}
+	a.callbacks = make([]logicalplan.Callback, 0, len(next.Callbacks()))
```

Comment:
It's confusing to me that the operators (here HashAggregate, but Distinction as well below) need to know about the multiple callbacks. I think that these operators should stay as they are (only aware of a single execution flow), but many copies of this operator should be instantiated at the physical planning stage for a single logical plan stage.

Reply:
Yes, you're right. 
What we really want should be a ConcurrentAggregator that's composed of one Aggregator, one Synchronizer, and n-Aggregators internally. 
I'll rework it to that, however, I think for the scope of just introducing multiple callbacks and for now only using the first callback, this should be a nice path into the component described above.

Reply:
I wouldn't make it so specific (i.e. there is no need for a new component ConcurrentAggregator). In the plan you described above, you would just pass the same synchronizer callback to each of the n aggregators (i.e. they will each call the synchronizer) and the synchronizer is the only component aware that it needs to handle concurrent calls to its callback method.